src/Optimus.php

```diff
@@ -93,12 +116,23 @@ public function decode($value)
             throw new InvalidArgumentException('Argument should be an integer');
         }
 
-        switch (static::$mode) {
+        $valXored = (int) $value ^ $this->xor;
+
+        $doMode = $this->mode;
+
+        if ($doMode === self::MODE_NATIVE
```

Comment:
Is this needed @courtney-miles?

Reply:
Do you mean as opposed to always using `native` or always `gmp`, rather than auto-switching to `gmp` on 64 bit systems only when dealing with the multiplication of large integers?

It could, in the constructor, force the mode to `gmp` when dealing with more than 31 bits for 64 bit numbers.

It was a speed decision to not use GMP if the specific values being multiplied do not need it. I was tempted to not do this because it makes the system more complicated.

I don't have the exact numbers now, but without GMP I was hashing X number of values in a couple of seconds, whilst with GMP that same number of values took 30 seconds.

This speed difference is not important to me and I'm happy to simplify and instead force GMP always when the bit length is greater than 31.

Reply:
I thought gmp was faster at handling big numbers.

Also; would it be possible to pass a random max int instead of a bit length? Eg. I want max int to be 10.000.000. Or will that not work with the maths?

Reply:
> I thought gmp was faster at handling big numbers.

It can handle bigger numbers than native PHP, but it seems if a number can be handled by native PHP then native PHP is faster. I'm happy to provide some timing if you like (I don't want to be responsible for unnecessary complexity.)

> Also; would it be possible to pass a random max int instead of a bit length?

In my tests max int must be a base 2 number to work. Please don't take my word for it -- I want to be wrong. So if you want to do some of your own tests, that would be great.

When researching this, I did find a similar strategy that uses a **coprime** in place of **prime** and **modulus** in place of **bitwise-and**.  That method seemed to allow any integer to be used as a max.  It is described at [A practical use of multiplicative inverses](https://ericlippert.com/2013/11/14/a-practical-use-of-multiplicative-inverses/)